packages/popcorn/src/utils/interpolate.ts

```diff
@@ -89,11 +94,21 @@ const slowInterpolate = <T>(
   };
 };
 
-export default <T>(
+function interpolate(
```

Comment:
The generics are more correct as it allows Typescript to infer what's going in by the arguments provided. So previously if you did this:

```
const a = interpolate([0,1], [0, 100]) // a is number
const b = interpolate([0,1], ['0%', '100%']) // b is string
```

Whereas this apporach, as far as I understand, leads to this:

```
const a = interpolate([0,1], [0, 100]) // a is number or string
const b = interpolate([0,1], ['0%', '100%']) // b is number or string
```

Reply:
From what I understand (Im baby stepping in TS, so Im not sure about it):
- previous version has allowed quite any generic T, it was not constrained to `string | number` and even if we would constrain it it would still allow `interpolate([0,1], [0, '100%'])`, I've tried various options, including conditional types and I couldnt make it work to accept either strictly `number[]` or `string[]` as output when using generics, I believe there should be an option to do this (especially when using conditional types, but Im not proficient enough to make it work)
- current version uses overloads which narrows the public API quite nicely, overloads before function declaration act as possible public APIs, whereas types in function declaration have to combine possible variants to handle the internal implementation, it's not a public API though, so even though w have 3 possible type definitions here, only 2 of them are public, and the inference works great for this - look here https://www.typescriptlang.org/play/#src=function%20interpolate(output%3A%20number%5B%5D)%3B%0D%0Afunction%20interpolate(output%3A%20string%5B%5D)%3B%0D%0Afunction%20interpolate(output%3A%20number%5B%5D%20%7C%20string%5B%5D)%20%7B%0D%0A%20%20%20%20return%20%5B%5D.concat(output)%20%20%0D%0A%7D%0D%0A%0D%0Aconst%20a%3A%20number%5B%5D%20%3D%20interpolate(%5B1%2C%202%5D)%0D%0Aconst%20b%3A%20string%5B%5D%20%3D%20interpolate(%5B'1'%2C%20'2'%5D)%0D%0Aconst%20c%3A%20(number%20%7C%20string)%5B%5D%20%3D%20interpolate(%5B1%2C%20'2'%5D)%0D%0A

Reply:
Ah I get it, thanks for the explanation. Awesome!